### DISCUSSION TITLE
Custom/uniform response schema

### DISCUSSION THREAD

**Discussion by johnbridstrup** [vitalik/django-ninja#1013] [Q&A]
I'd like to have a custom response format for all responses in my api, something like
```json
{
  "data": {...}, # What would normally be returned
  "code": 200,
  "meta": "info"
}
```

Is there any way to do this built in? If not, would I want to create a custom `Schema` or `ModelSchema` class to do this? Not sure where to begin so figured I'd ask here first.

**Comment by johnbridstrup** [comment]
I considered a decorator along the lines of
```python
def wrap_response(schema):
  class FullSchema(Schema):
    data: schema
    code: int
    meta: str
  return FullSchema
```
But then I would have to do extra work building the response in the endpoints. I'd still like to be able to just return e.g. `200, query`

edit: it also breaks with list responses of a schema

**Comment by vitalik** [comment]
@johnbridstrup 

could you show some use cases ? would you fill "code=200" inside view function or it should just copy http response status ? same for meta=info

**Reply by johnbridstrup** [reply]
Thanks for responding! I suppose either way works, I'm not sure what is easier or more optimal (quite new to ninja). I'm actually kind of surprised this functionality isn't common in the frameworks with openapi integrated... for instance in a DRF project at work we wanted a uniform response structure

```json
{
  "data": {},
  "message": "A message",
  "status": 200
}
```

We did this with custom json renderer and autoschema classes. We then had `make_ok` and `make_error` functions to create the dict that was eventually rendered. The reason was mostly for provide the message field for additional context about each request, if needed, while still making use of `ModelViewSet` and all the nice automatic schema/view generation.  An important aspect of how we did it here is that pagination does not break, because all of the modification happens at the very end (after schema creation and right when the response is rendered).

I'm able to do the response structure in a similar way thanks to #445, its the schema part I'm struggling with. I would really prefer the schema to match the actual response. Hope this clarifies!

**Reply by vitalik** [reply]
ok.. but could you provide some close to real life example 

let's say you have a hello method:

```Python

class Hello(Schema):
      name: str

@api.get('/hello/{name}', response=Hello)
def hello(request, name: str):
     return {"name", "Hello"}

```

how in this case you would want to override your "status" and "info" ? or will you want to override it ?

**Reply by johnbridstrup** [reply]
I think in that case I'd want to

```python
return code, {"name": "hello"}, {**extra_info}
```

Where `extra_info` is more like a named tuple, then unpack it in `create_response`.

**Reply by johnbridstrup** [reply]
I'm using model schemas mostly, so typically I'm returning querysets, if that's relevant.
